src/wxVolumeRenderer.cpp

```diff
@@ -0,0 +1,361 @@
+//
+//  wxVolumeRenderer.cpp
+//  ballpit
+//
+//  Created by Benedikt Stefan Vogler on 20.08.20.
+//
+
+#include <wx/wxprec.h>
+
+#ifndef WX_PRECOMP
+#  include <wx/wx.h>
+#endif
+
+#ifdef __APPLE__
+#include <OpenCL/opencl.h>
+#else
+#include <CL/cl.h>
+#endif
+
+#include <chrono>
+#include <iostream>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "controller.h"
+#include "model.h"
+#include "wxVolumeRenderer.h"
+#include <wx/rawbmp.h>
+
+BEGIN_EVENT_TABLE(wxVolumeRenderer, wxPanel)
+
+// catch paint events
+EVT_PAINT(wxVolumeRenderer::paintEvent)
+
+END_EVENT_TABLE()
+
+
+wxVolumeRenderer::wxVolumeRenderer(wxFrame* parent,
+						   wxBitmapType format) : wxPanel(parent){
+	auto width = 512;
+	auto height = 512;
+	unsigned char* colormatrix = createImageGPU("../res/renderkernel.cl", width, height);
```

Comment:
I compiled this branch version on windows 10 but the command prompt displays this message when trying to render the voxels "Failed to load kernel ../res/renderkernel.cl."
Does the file renderkernel.cl need to be included in some way during compilation? Is there a specific way to deal with these .cl files in makefile?

Reply:
It seems that on windows, the relative path should start with a single dot.
I had  to change the line 
`unsigned char* colormatrix = createImageGPU("../res/renderkernel.cl", width, height);`
to
`unsigned char* colormatrix = createImageGPU("./res/renderkernel.cl", width, height);`
for the program to find the file.

But now I get another error: `Error: Failed to allocate image! Error: -40`
I will investigate.

Reply:
Well, after looking further in the code, it is a custom error message from line 208. The error code -40 however is specific to OpenCl and indicates that the image size is invalid (see https://streamhpc.com/blog/2013-04-28/opencl-error-codes/ ).
After looking further, it seems that the issue is coming from the values of `image_row_pitch` and/or `image_slice_pitch` from lines 195 and 196.
I tried to change manually the values following info found in this link http://man.hubwiz.com/docset/OpenCL.docset/Contents/Resources/Documents/cl_image_desc.html but only value 0 overcame the error. Unfortunately, it leads to other errors further.
I am not knowledgeable enough in OpenCl to fix these issues so I will let you look into it.

Reply:
> Found the issue. It seems that we don't generate the executable in the same folder (for me it is in bin/release), so the relative path was not working for me.

AFAIK the makefile does not create a debug/release distinction. If you run a custom compilation/linking script, it can not be supported. 

> Now with the files in the right folders, I get another error

Interesting. On my machine, it compiles, links, and runs fine in the executable. What is your hardware? Are you running dedicated GPU or integrated graphics?
It seems that I also have to test it on my Windows machine.
I had this strange issue that the OpenCL compiler at some point failed at compiling the OpenCL code when compiling outside the runtime. `float4(0.0)` is a valid call.

Reply:
I think one remaining issue stems from the assumption that the number of chunks is uniform in each dimension. I only tested with `isobutane.xyz` and `acetylene.xyz` because these were the only available files to me. The recently included `6s8y.xyz` crashes on the matrix generation step.

Reply:
Please try again, as I have fixed the dimension issues. Note that on some new input files the octaves to matrix conversion can take several seconds. The matrix is 160 MB. There should be probably some strategies in the future to sample at a lower rate or generate the image in a second thread. I could also imagine continuous rendering and filling of the matrix.

Reply:
> AFAIK the makefile does not create a debug/release distinction. If you run a custom compilation/linking script, it can not be supported.

Yes, I don't use the makefile from this repository as I use code::blocks and manage libs manually. For this branch, I only had to add an opencl lib (I put the one from NVidia cuda toolkit `cuda_11.0.3_451.82_win10` ) to compile successfully.

> Interesting. On my machine, it compiles, links, and runs fine in the executable. What is your hardware? Are you running dedicated GPU or integrated graphics?

I have both Intel(R) UHD Graphics 630 and NVIDIA GeForce GTX 1660 Ti on my laptop. I tried to run the program with both and they lead to the same errors (only during runtime when calling the voxel renderer).

> The recently included 6s8y.xyz crashes on the matrix generation step.

This is a protein structure which is very large. So the size of the image in imageDescriptor will probably be set too high (above CL_DEVICE_IMAGE3D_MAX_WIDTH, CL_DEVICE_IMAGE3D_MAX_HEIGHT, CL_DEVICE_IMAGE3D_MAX_DEPTH).
I recently added smaller structures in the inputfile folder (`fullerene C60.xyz` and `Pd6L4_open_cage_Fujita.xyz`) which should work fine for you.
In any case, I intend to export the surfaces as map files readable on other chemical structure rendering softwares (PyMol and Chimera) so there will be an alternative if your renderer can't display such large protein structures.

> Please try again, as I have fixed the dimension issues.

The command prompt displayed these messages:
```
Time for octrees to matrix conversion 11.168 ms
Error: Failed to allocate image! Error: -40
```

Reply:
it can now render every included structure file. Maybe @jmaglic can try on his windows set-up.